resource_impl.h

```diff
@@ -23,26 +23,106 @@
 #include <errno.h>
 
 #define RESOURCE_64	64
+#define RESOURCE_256	256
 #define RESOURCE_2048	2048
 
+#define FNAMELEN	RESOURCE_256
+
+#define INITSCOPE	"/init.scope"
+
+/* We are assuming that cgroup is mounted at its usual location,
+ * If we have reeust from applications who do not mount cgroup at
+ * usual location, then we need to change this.
+ */
+#define DEFAULTCGFS	"/sys/fs/cgroup"
+
 #define eprintf(msg, ...)	fprintf(stderr,\
 					"Err at line %d in file %s: "msg"\n",\
 					__LINE__, __FILE__, ##__VA_ARGS__)\
 
+static inline void clean_init(char *cg)
+{
+	char *p;
+	size_t cg_len = strlen(cg), init_len = strlen(INITSCOPE);
+
+	if (cg_len < init_len)
+		return;
+
+	p = cg + cg_len - init_len;
+	if (strcmp(p, INITSCOPE) == 0) {
+		if (p == cg)
+			*(p+1) = '\0';
+		else
+			*p = '\0';
+	}
+}
+
+/* Get cgroup path for a particular controller */
+static inline char *get_cgroup(pid_t pid, const char *contrl)
+{
+	char fn[FNAMELEN];
+	FILE *f;
+	char *line = NULL;
+	size_t len = 0;
+	size_t l = 0;
+	int ret;
+	char *cgrp = NULL;
+	char *c1, *c2;
+
+	/* If no pid is provided then return cgroup info for current process.
+	 */
+	if (pid) {
+		ret = snprintf(fn, FNAMELEN, "/proc/%d/cgroup", pid);
+	} else {
+		ret = snprintf(fn, FNAMELEN, "/proc/self/cgroup");
+	}
+
+	if (ret < 0 || ret >= FNAMELEN)
+		return NULL;
+
+	if (!(f = fopen(fn, "r")))
+		return NULL;
+
+	while (getline(&line, &len, f) != -1) {
+		if (!line[0])
+			continue;
+		c1 = strchr(line, ':');
+		if (!c1)
+			goto out;
+		c1++;
+		c2 = strchr(c1, ':');
+		if (!c2)
+			goto out;
+		*c2 = '\0';
+		if (strcmp(c1, contrl) != 0)
+			continue;
+		c2++;
+		l = strlen(c2);
+		if (l && c2[l-1] == '\n')
+			c2[l-1] = '\0';
+		if(strcmp(c2, "/") == 0)
+			goto out;
+		do {
+			cgrp = strdup(c2);
+		} while (!cgrp);
```

Comment:
I would just do a simple `cgrp = strdup(c2)`. I once was in favor of the looping approach but I come to like it less and less because really this should only ever properly fail if one runs out of address space or sm which is the end of everything anyway.

Reply:
Ok.